src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-	 * value in s0. */
-	unsigned index = 2;
-	while (index < count) {
-		riscv_addr_t read_addr = address + index * increment;
-		LOG_DEBUG("i=%d, count=%d, read_addr=0x%" PRIx64, index, count, read_addr);
-		/* The pipeline looks like this:
-		 * memory -> s1 -> dm_data0 -> debugger
-		 * Right now:
-		 * s0 contains read_addr
-		 * s1 contains mem[read_addr-size]
-		 * dm_data0 contains[read_addr-size*2]
-		 */
+/**
+ * This function attempts to restore the pipeline after a busy on abstract
+ * access.
+ * Target's state is as follows:
+ * s0 contains address + index_on_target * increment
+ * s1 contains mem[address + (index_on_target - 1) * increment]
+ * dm_data[0:1] contains mem[address + (index_on_target - 2) * increment]
+ */
+static int read_memory_progbuf_inner_on_ac_busy(struct target *target,
+		uint32_t start_index, uint32_t elements_to_read, uint32_t *elements_read,
+		struct memory_access_info access)
+{
+	increase_ac_busy_delay(target);
+	riscv013_clear_abstract_error(target);
 
-		struct riscv_batch *batch = riscv_batch_alloc(target, RISCV_BATCH_ALLOC_SIZE,
-				info->dmi_busy_delay + info->ac_busy_delay);
-		if (!batch)
+	if (dmi_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* See how far we got by reading s0/a0 */
+	uint32_t index_on_target;
+
+	if (/*is_repeated_read*/ access.increment == 0) {
+		/* s0 is constant, a0 is incremented by one each execution */
+		riscv_reg_t counter;
+
+		if (register_read_direct(target, &counter, GDB_REGNO_A0) != ERROR_OK)
 			return ERROR_FAIL;
+		index_on_target = counter;
+	} else {
+		target_addr_t address_on_target;
 
-		unsigned reads = 0;
-		for (unsigned j = index; j < count; j++) {
-			if (size > 4)
-				riscv_batch_add_dmi_read(batch, DM_DATA1);
-			riscv_batch_add_dmi_read(batch, DM_DATA0);
+		if (register_read_direct(target, &address_on_target, GDB_REGNO_S0) != ERROR_OK)
+			return ERROR_FAIL;
+		index_on_target = (address_on_target - access.target_address) /
+			access.increment;
+	}
 
-			reads++;
-			if (riscv_batch_full(batch))
-				break;
-		}
+	/* According to the spec, if an abstract command fails, one can't make any
+	 * assumptions about dm_data registers, so all the values in the pipeline
+	 * are clobbered now and need to be reread.
+	 */
+	const uint32_t min_index_on_target = start_index + 2;
+	if (index_on_target < min_index_on_target) {
+		LOG_TARGET_ERROR(target, "Arithmetic does not work correctly on the target");
+		return ERROR_FAIL;
+	} else if (index_on_target == min_index_on_target) {
+		LOG_TARGET_DEBUG(target, "No forward progress");
+	}
+	const uint32_t next_index = (index_on_target - 2);
+	*elements_read = next_index - start_index;
+	LOG_TARGET_WARNING(target, "re-reading memory from addresses 0x%"
+			TARGET_PRIxADDR " and 0x%" TARGET_PRIxADDR,
+			access.target_address + access.increment * next_index,
+			access.target_address + access.increment * (next_index + 1));
```

Comment:
I was surprised to see that you mention 2 addresses here. I think that is because read_memory_progbuf_inner_startup() already does 2 reads, so by the time you get here there have always been 2 reads already. Is that right?

Reply:
User-visible messages should generally be capitalized and have punctuation like regular English sentences. (Note that this breaks your riscv-tests change. Simplest is to make that one ignore capitalization.)

Reply:
> User-visible messages should generally be capitalized and have punctuation like regular English sentences. (Note that this breaks your riscv-tests change. Simplest is to make that one ignore capitalization.)

Addressed.

> I was surprised to see that you mention 2 addresses here. I think that is because read_memory_progbuf_inner_startup() already does 2 reads, so by the time you get here there have always been 2 reads already. Is that right?

Yes, that's right.
At all times in the loop, there are two values loaded from memory, but not yet read by the debugger. The first value is in `s1` and the other in `dm_data[0:1]`.
If the abstract command fails, both values can get clobbered, so it is necessary to re-read them.